### DISCUSSION TITLE
Help in running the code on my esp32

### DISCUSSION THREAD

**Discussion by ayazalavi** [paullouisageneau/libdatachannel#675] [General]
I am using esp idf to develop my project. I downloaded libdatachannel in esp folder so now I have 2 directories in esp folder:

1. esp-idf
2. libdatachannel

I then updated CMakeLists.txt in the root of my project and added line

set(EXTRA_COMPONENT_DIRS "$ENV{IDF_PATH}/examples/common_components/protocol_examples_common $ENV{IDF_PATH}/../libdatachannel")

now when I try to build the project I get following callstack

Executing task: ninja  

[0/1] Re-running CMake...
-- Project is not inside a git repository, or git repository has no commits; will not use 'git describe' to determine PROJECT_VER.
-- Building ESP-IDF components for target esp32
CMake Error at /Users/ayaz/esp/esp-idf/tools/cmake/component.cmake:219 (message):
  CMake Warning (dev) at build_properties.temp.cmake:7:

    Syntax Warning in cmake code at column 51

    Argument not separated from preceding token by whitespace.

  Call Stack (most recent call first):

    /Users/ayaz/esp/esp-idf/tools/cmake/scripts/component_get_requirements.cmake:3 (include)

  This warning is for project developers.  Use -Wno-dev to suppress it.

  fatal: not a git repository (or any parent up to mount point /Volumes)

  Stopping at filesystem boundary (GIT_DISCOVERY_ACROSS_FILESYSTEM not set).

  CMake Error at /Users/ayaz/esp/libdatachannel/CMakeLists.txt:2 (project):

    project command is not scriptable

  Call Stack (most recent call first):

    /Users/ayaz/esp/esp-idf/tools/cmake/scripts/component_get_requirements.cmake:106 (include)
    /Users/ayaz/esp/esp-idf/tools/cmake/scripts/component_get_requirements.cmake:124 (__component_get_requirements)

Call Stack (most recent call first):
  /Users/ayaz/esp/esp-idf/tools/cmake/build.cmake:491 (__component_get_requirements)
  /Users/ayaz/esp/esp-idf/tools/cmake/project.cmake:378 (idf_build_process)
  CMakeLists.txt:10 (project)

-- Configuring incomplete, errors occurred!
See also "/Volumes/Projects/ESP-Projects/websocket/build/CMakeFiles/CMakeOutput.log".
FAILED: build.ninja 
/Users/ayaz/.espressif/tools/cmake/3.20.3/CMake.app/Contents/bin/cmake --regenerate-during-build -S/Volumes/Projects/ESP-Projects/websocket -B/Volumes/Projects/ESP-Projects/websocket/build
ninja: error: rebuilding 'build.ninja': subcommand failed

 *  The terminal process "/bin/zsh '-c', 'ninja '" terminated with exit code: 1. 

I need help in fixing this problem

EDITED:
So far I have been able to build libdatachannel for Mac and updated my CMakeLists.txt file to below:
```
cmake_minimum_required(VERSION 3.5)

# (Not part of the boilerplate)
# This example uses an extra component for common functions such as Wi-Fi and Ethernet connection.
set(EXTRA_COMPONENT_DIRS $ENV{IDF_PATH}/examples/common_components/protocol_examples_common)

include($ENV{IDF_PATH}/../libdatachannel/build/LibDataChannelConfigVersion.cmake)
include($ENV{IDF_PATH}/tools/cmake/project.cmake)
project(websocket_example)
```

**Comment by paullouisageneau** [comment]
I'm not very familiar with ESP32, but it seems to me that the ESP32 build system requires specifically-written CMakeLists files to register components, you can't just include a normal CMakeLists (see [the documentation](https://docs.espressif.com/projects/esp-idf/en/release-v4.1/api-guides/build-system.html#example-component-cmakelists)).

Additionally, ESP-IDF is based on the real-time operating system FreeRTOS, while libdatachannel requires a POSIX-compatible operating system, therefore building the library might require heavy patching.

**Reply by ayazalavi** [reply]
Can we generate .a file for libdatachannel?

**Reply by paullouisageneau** [reply]
The target `datachannel-static` will build a couple of ".a" static libraries:
```
$ cmake -B build -DCMAKE_BUILD_TYPE=Release
$ cd build
$ make datachannel-static
$ find -type f -name "*.a"
```
However, you can't just link static libraries built for your computer, you need them compiled for the ESP32.
